src/server/pfs/fuse/server.go

```diff
@@ -152,90 +172,53 @@ func (mm *MountManager) ListByRepos() (ListRepoResponse, error) {
 				rr.Authorization = "read"
 			}
 		}
-		lr[repo.Repo.Name] = rr
+		lr = append(lr, rr)
 	}
 
-	// TODO: also add any repos/branches that have been deleted from pachyderm
-	// but are still mounted here :-O we should send them a special signal to
-	// tell them they're "stranded" or "missing" probably?
-
 	return lr, nil
 }
 
-func (mm *MountManager) ListByMounts() (ListMountResponse, error) {
+func (mm *MountManager) ListByMounts(projectFilter string) (ListMountResponse, error) {
 	mm.mu.Lock()
 	defer mm.mu.Unlock()
 
 	mr := ListMountResponse{
-		Mounted:   map[string]MountState{},
-		Unmounted: map[string]RepoResponse{},
-	}
-	// Keep track of existing repos and branches to see if mm.States contains deleted repos/branches
-	repoBranches := map[string]map[string]bool{}
-	repos, err := mm.Client.ListRepo()
-	if err != nil {
-		return mr, err
-	}
-	for _, repo := range repos {
-		repoBranches[repo.Repo.Name] = map[string]bool{}
-		rr := RepoResponse{Repo: repo.Repo.Name}
-		readAccess := true
-		if repo.AuthInfo != nil {
-			readAccess = hasRepoRead(repo.AuthInfo.Permissions)
-			rr.Authorization = "none"
-		}
-		if readAccess {
-			bis, err := mm.Client.ListProjectBranch(repo.Repo.Project.GetName(), repo.Repo.Name)
-			if err != nil {
-				// Repo was deleted between ListRepo and ListBranch RPCs
-				if auth.IsErrNoRoleBinding(err) {
-					continue
-				}
-				return mr, err
-			}
-			for _, bi := range bis {
-				repoBranches[repo.Repo.Name][bi.Branch.Name] = true
-				rr.Branches = append(rr.Branches, bi.Branch.Name)
-			}
-			if repo.AuthInfo == nil {
-				rr.Authorization = "off"
-			} else if hasRepoWrite(repo.AuthInfo.Permissions) {
-				rr.Authorization = "write"
-			} else {
-				rr.Authorization = "read"
-			}
-		}
-		mr.Unmounted[repo.Repo.Name] = rr
+		Mounted:   []MountState{},
+		Unmounted: []RepoResponse{},
 	}
 
-	// Unmount mounted repos/branches that were deleted
 	for name, msm := range mm.States {
 		if msm.State == "mounted" {
-			_, ok := repoBranches[msm.Repo]
-			if !ok {
-				mm.unmountDeletedRepos(name)
-				continue
-			}
-			_, ok = repoBranches[msm.Repo][msm.Branch]
-			if !ok && msm.Mode == "ro" {
-				mm.unmountDeletedRepos(name)
-				continue
+			// Check if a mounted repo/branch was deleted to remove it from state.
+			// In read-only mode, a branch must exist to mount. In read-write mode, it
+			// is not necessary for the branch to exist, as a new one
+			if msm.Mode == "ro" {
+				if exists, _ := mm.verifyProjectRepoBranchExist(msm.Project, msm.Repo, msm.Branch); !exists {
```

Comment:
Wondering if at some point it might make sense to combine some of this logic into a single VerifyExistence(msm.Mode, msm.Project, msm.Repo, msm.Branch) call.

Reply:
Added